### PR TITLE
Apply system_default configuration on SSL_CTX_new()

### PR DESCRIPTION
**Pull request by t8m**
When SSL_CTX is created preinitialize it with system default
configuration from system_default section.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by kroeckx** [comment]
I'm wondering why at SSL_CTX_new, and not when the library is loaded? Do you think there is an advantage to do it at that time? Like that daemons create a new SSL_CTX_new() on SIGHUP?

**Comment by t8m** [comment]
The file is actually loaded on OPENSSL_init_ssl() so the commit message is not correct, I'll modify it. However it would actually be nice if it was possible to re-load the file on SSL_CTX_new() otherwise it would require restart of all daemons to apply the changed system policy. I wonder if there is a possibility to reload the configuration file without unwanted side-effects.

**Comment by nmav** [comment]
I've opened issue #5041 documenting what this merge request addresses.

**Comment by t8m** [comment]
Test and documentation added. Can you please still consider this for 1.1.1?

**Comment by richsalz** [comment]
Sigh.  Nobody else on the team seems to care much about this, unfortunately.  I care a lot but I am only one person.  Ping @openssl/committers

**Comment by kroeckx** [comment]
I've always said that I was interrested in this. I will try to
find some time to review this if nobody else does.

**Comment by vdukhovni** [comment]
One risk is that this can introduce security issues in Postfix by unexpectedly enabling trust in the default CAs (CAfile and CApath) in the SMTP server, which till now would by default trust only specific CAs (perhaps just a local one) and possibly allow relaying of email from all clients that have a client from such a CA.

If the trust is expanded behind Postfix's back to include all the WebPKI CAs, the system becomes an open relay.

There may other costs and downside to untroducing this as a new behaviour of SSL_CTX_new().  It might make more sense to encourage applications that want this type of system-wide configuration to use a new function that has the desired behaviour. SSL_CTX_new_default() or similar.

**Comment by t8m** [comment]
What about just filtering out / ignoring the certificate related options from the system-default?

I mean by adding a new API and requiring applications to switch to it, we can dump this altogether as this is not going to be "system default" but "some changed applications default".

Although I know that even the current patch is not ideal in this regard either because it requires the application to initialize the OpenSSL with OPENSSL_init_ssl(OPENSSL_INIT_LOAD_CONFIG) otherwise the configuration file is not loaded. But at least patching OpenSSL initialization in applications should be much more easily achievable than patching all calls to SSL_CTX_new().

**Comment by p-steuer** [comment]
Id support the general idea of providing a system-wide configuration for libssl and libcrypto. For example it should be possible to enable platform specific hardware support via engine on system level rather than requiring applications to be aware of the hw they are running on. The same applies to entropy source configuration, see my comment in #4394 .

This is not directly related to this specific PR, but there are multiple PRs/issues regarding configuration file and discussion seems to take place here ..

**Comment by richsalz** [comment]
I am glad to finally see some discussion on this.  I think the expectation is that this would be used to increase the default level of security for all applications. But Viktor raises a good point about things like changing the default trust store affecting some applications in the wrong way.  I think those applications are in the minority however.

Perhaps instead of having an "opt-in" we have an "opt-out" and applications like postfix would use it.

**Comment by vdukhovni** [comment]
Either opt-out or a new API entry point (opt-in) is I think required.  I'm also concerned about the performance impact of whatever processing this might imply.  SSL_CTX_new() is called not only when the application first initializes, but IIRC also when creating contexts for SNI, and this should involve excessive processing that can be avoided, or unwanted processing that might bring in unexpected settings.

My general view is that OpenSSL provides low-level primitives that can be composed to yield higher-level user-friendly functions.  This makes it possible to make various choices, without being forced into a one-size fits all model.  The downside is that we've not always (or often not) provided the higher-level user-friendly interfaces.  So we should do more of that, but not necessarily by converting the low-level functions to one-size fits all common-case interfaces.

Thoughts?

**Comment by kroeckx** [comment]
I want to have defaults in Debian that are more strict than the
upstream defaults. I would like to move those defaults to some
config file. I want users to be able to edit that config file, and
I want applications be able to override the defaults.

**Comment by FdaSilvaYY** [comment]
Either opt-in or opt-out, this feature betterbe visible/displayed when calling a kind of `openssl version -a'` command.

**Comment by nmav** [comment]
My understanding is that @paulidale @richsalz and @kroeckx are ok with that summary (please correct me if I'm wrong). @vdukhovni is that summary ok with you? Do you have any suggestion for the opt-out mechanism? Is not calling `OPENSSL_init_ssl(OPENSSL_INIT_LOAD_CONFIG)` ok or something else should be introduced?

**Comment by bernd-edlinger** [comment]
I can't see why an application should not be able to create a blank context.

**Comment by kaduk** [comment]
The above summary does not explicitly take into account a transition plan from the current state to the "new" state of affairs.

**Comment by vdukhovni** [comment]
The summary is close, but I don't think that opt-out is the right way to introduce a higher-level API that imports system-wide settings. For that, I think it makes more sense to introduce new functions, leaving SSL_CTX_new() as a basic building block.

In addition to being able to import system-wide settings, a newer API could also facilitate automatic reloading of private keys and the associated certificate chain transparently when the containing files are modified.  Having to restart long-running applications when the keys rotate is a problem to avoid. Certificate lifetimes are getting shorter, and not all applications are able to restart without interrupting service. There are probably various other features we'd want, so I think it is time to do something new high-level rather than retrofit.

**Comment by bernd-edlinger** [comment]
+1

**Comment by kroeckx** [comment]
We have over 600 source packages in Debian making use of OpenSSL.
I do not want to have to look at all of them to see if they need a
patch, and provide a patch for probably half of them to be able to
set defaults. I would rather patch the 3 exceptions.

**Comment by nmav** [comment]
> The above summary does not explicitly take into account a transition plan from the current state to the "new" state of affairs.

@kaduk I am not sure what do you mean by the transition, but the feature is a no-op unless someone (administrator or distributor) modifies the configuration file. The intention at least from our view, is that this pull request is to provide the back-end in openssl; distributors (like Debian and Fedora) will provide a transition plan if needed.

> I can't see why an application should not be able to create a blank context.

@bernd-edlinger isn't that covered by "Applications must have the option to opt-out from system-wide defaults", or are you referring to something else?

> The summary is close, but I don't think that opt-out is the right way to introduce a higher-level API that imports system-wide settings. 

@vdukhovni this PR introduces a way to set default settings system-wide.  Such default settings have value when they are default to all applications. Otherwise as @kroeckx and @t8m  already mentioned the effort to modify each and every app is prohibitive. Note also that this PR only changes defaults if the distributor of the software has modified the config file, not otherwise.

**Comment by bernd-edlinger** [comment]
> isn't that covered by "Applications must have the option to opt-out from system-wide defaults", or are you referring to something else?

Your plan looks like an all-or-nothing thing.
And this PR in its current state looks pretty much in-acceptable to me.

I would like to have a layer of configurable SSL functions on top of a
layer basic SSL functions.  Not everything munged into one function.

What SSL_CTX_new does now must still be available, without any
performance impact, like parsing a config file each time a new context
is created.  It is of course possible that SSL_CTX_new gets renamed
to something different, so that 600 unmaintained packages can pick the
auto-configured function by the name SSL_CTX_new for instance.

When I want an SSL to use system defaults, there must be a function
to accomplish that task, but when I want to setup everything bit
for bit, there must also be a way to do that.
When I want an application-specific configuration file I want a
function to accomplish that task as well.

**Comment by vdukhovni** [comment]
It is not clear that Debian should be opting upstream packages into new, unexpected, OpenSSL behaviour.  I don't think that's appropriate, certainly not without porting the package and coordinating with the upstream maintainers.

**Comment by kaduk** [comment]
> @kaduk I am not sure what do you mean by the transition, but the feature is a no-op unless someone (administrator or distributor) modifies the configuration file. The intention at least from our view, is that this pull request is to provide the back-end in openssl; distributors (like Debian and Fedora) will provide a transition plan.

The quoted text is the sort of thing that I was looking for  (in this case, that the new semantics take effect silently at library update time, but functionality change occurs upon action by the system administrator).

This got some other comments already (sorry for the slow response), but I am inclined to also say that this transition plan is not a good one, as the risk of "silent" (i.e., not loud-and-obvious at library-upgrade time) breakage is too high.  We may be stuck adding a new "high-level" API that wraps SSL_CTX_new() and the additional global configuration stuff, and try to get consumers to switch.  Changing the functionality of SSL_CTX_new() itself could probably only be done on a major release boundary (i.e., 1.2.0), if it is worth doing at all.

**Comment by nmav** [comment]
> but I am inclined to also say that this transition plan is not a good one, as the risk of "silent" (i.e., not loud-and-obvious at library-upgrade time) breakage is too high. We may be stuck adding a new "high-level" API that wraps SSL_CTX_new() and the additional global configuration stuff, and try to get consumers to switch. Changing the functionality of SSL_CTX_new() itself could probably only be done on a major release boundary (i.e., 1.2.0), if it is worth doing at all.

@kaduk I have no opinion on that, the release schedule is up to openssl devs.

> It is not clear that Debian should be opting upstream packages into new, unexpected, OpenSSL behaviour. I don't think that's appropriate, certainly not without porting the package and coordinating with the upstream maintainers.

@vdukhovni That's certainly a respectable opinion, but I believe it should be up to OS maintainers to decide whether they want to follow your advice. Tomas' proposal doesn't come out of the blue. The approach of applying system-wide settings was tested in the majority of TLS using packages since [Fedora 21](https://fedoraproject.org/wiki/Changes/CryptoPolicy), and in all openssl packages [at Red Hat Enterprise Linux 6.9](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/6.9_technical_notes/chap-red_hat_enterprise_linux-6.9_technical_notes-deprecated_functionality). The RHEL change disables insecure signature algorithms, from openssl while the Fedora change is restricting the ciphersuites presented by openssl to sets which are consistent with all the other system libraries (nss, gnutls, openjdk, etc). For the Fedora change [we went through all packages using openssl](https://bugzilla.redhat.com/show_bug.cgi?id=1179209) (as well as to other implementations), to ensure their fitness into the system-wide settings. Of these only postfix (which I believe you represent), and one other application didn't fit into the defaults (it wasn't using openssl and was using anonymous ciphersuites). Since 2014 which the Fedora change took place, we have zero issues reported by users of Fedora or customers of ours related to that change.

So, while I understand your concerns, but I do not believe they apply to the vast majority of applications shipped in distributions. What I would like to ask you (and any other) is to help us constructively design an interface which can help both distributors and application developers achieve our goals realistically.

**Comment by t-j-h** [comment]
I agree that this is missing functionality and the opt-out approach is the only practical way to achieve the desired objective - downstream have done this for a reason - and it is (in my view) the behaviour that the average user will want - to be able to apply defaults to all applications without changing the code.

One possible way to sort this out is to have the opt-out approach also work via this mechanism - and in a linux and windows (and many other unix-like systems) we could easily do that by looking at the application name and having a list of applications which should not be impacted by the system-wide defaults. That way you wouldn't even have to patch those either.

An API to allow ignoring the system defaults is also reasonable to add for the few applications that will actually want to ignore such things (and frankly I don't think they are many of those and I'd argue that those applications are likely to end up replicating the configuration file settings update approach too).

So in summary - I think we should be doing this - but with the added ability to bypass it on a per-named  application basis.

**Comment by richsalz** [comment]
I think that the trade-off of "distro's and systems" versus one particular security-aware application, basically *has* to come down on the side of the first group.  While we haven't deliberately tried to inconvenience downstream, it has been awhile since we have done something that particular meets a need of theirs.

postfix should be able to opt-out by explicitly initializing openssl or some other mechanism.  Everyone else will benefit by free security upgrades.

**Comment by vdukhovni** [comment]
What does the proposed approach do with secondary SSL_CTX objects for SNI?  What is the overhead of the additional initialization?  What is the API for opting out?

The historical behaviour of SSL_CTX_new() is to allocate a new SSL_CTX object and do nothing else.  How is that behaviour to be obtained if the contract is changed?  However well motivated, fundamentally new behaviours should require a new function, rather than be retrofitted onto existing functions.

If the distro wants to have all applications using the default stores, then they can build those applications with a #define to make it so.

**Comment by kroeckx** [comment]
There seem to be a lot of talk about SSL_CTX_new() here. I would
rather have it that we read the config file at library
initialization. If changes on the fly need to be supported we can
think about that later.

**Comment by bernd-edlinger** [comment]
This PR changes the behaviour of SSL_CTX_new with no way to access the original behaviour.
Why can't we not get the original function, maybe as SSL_CTX_new_unconfigured ?
Additionally to opt-out at initialization time?

**Comment by vdukhovni** [comment]
Whatever we do, "looking at the application name" is not a robust mechanism we should consider.

**Comment by kroeckx** [comment]
This is really just a pull request to start the discussion, please
ignore whatever the patch currently does.

**Comment by t8m** [comment]
Please note that no config files will be read on SSL_CTX_new with the patch. Just the configuration from the file read on initialization is applied to the SSL_CTX. I do not really think performance would be an issue for large majority of OpenSSL using apps.

OK, so what about this proposal:
Add SSL_CTX_new_ex() with additional parameter flags which would allow to initialize the SSL_CTX with or without applying the system default settings.
As for whether SSL_CTX_new() should call the SSL_CTX_new_ex() with the flag to apply the system default or not to apply to them, we could leave it to later discussion.

If the #2397 lands first it should already add the flags parameter so we do not have to create additional API call.

**Comment by t8m** [comment]
@vdukhovni the distros already handle default CA stores somehow. Here the major driver is a crypto policy support in regards to enabled algorithms and minimum key size limits.

**Comment by levitte** [comment]
@richsalz, you removed the "Post 1.1.1" milestone...  what milestone would you like to see?

**Comment by richsalz** [comment]
I want this in 1.1.1  I want us to make it easy for admins and *downstream distro maintainers* to change defaults, even if a postfix maintainer doesn't like it.

**Comment by levitte** [comment]
Ok, so 1.1.1 milestone it is

**Comment by t-j-h** [comment]
I'm less concerned about change-on-the-fly approaches - i.e. just having a one-time load of the configuration file at startup and applying that would meet my expectations. Checking the configuration file and reloading it for every call to SSL_CTX_new isn't what I would expect - and if that was done it should be on a timestamp check caching basis at least. And be able to be overridden - which I guess argues for that being a configuration option - loaded at startup - should we be checking for configuration changes for each SSL_CTX_new call - as a suggestion for how to handle it.

And I also think this should be going into the 1.1.1 release as well ...

**Comment by richsalz** [comment]
I think the simplest thing to understand is read the file at startup, don't check for run-time changes.

**Comment by t8m** [comment]
@richsalz Which is what actually happens with the current patch. So do I understand it correctly that the only missing thing (if vote passes) is adding the opt-out interface? I'd say it should also accommodate the #3761 to avoid proliferation of new API calls.

**Comment by richsalz** [comment]
Yes, that's my opinion.  There is some discussion going on in openssl-project mailing list.

**Comment by richsalz** [comment]
Merged.  Please see #5671 for some remaining issues.
